# Data Handling with Pandas

[Pandas](https://numpy.org/) Is the most popular open source tool for data analysis and manipulation, built on top of the Python programming language.

#### Learning objectives
 By the end of this notebook you will know:
+ How to use Pandas to read and write data files
+ How to get some basic information about a Pandas DataFrame
+ What are subsets and how to create them
+ How to create a confusion matrix
---

## **1. Read data files with Pandas**:
* Dataframes are the main data structure in Pandas.
* Dataframes are two dimensional tables whose columns have names and potentially store different types of data
* Load the library with `import pandas as pd`. The alias pd is commonly used for Pandas. (recommended)
* Read a Comma Separated Values (CSV) data file with `pd.read_csv`.
  - The main paramenter is the URL or name of the file to be read.
  - Assign result to a variable to store the data that was read.

In [1]:
# import the library
import pandas as pd

# directory to the file, it can be from own computer or from web
url = 'https://raw.githubusercontent.com/MaastrichtU-IDS/global-studies/main/semester2/data/gapminder.csv'

# read the url and save it as a variable called df
df = pd.read_csv(url)

- Normally, as a convention, people use `df` as the standard name to call the data, df stands for DataFrame, indicating that one should not confuse this data structure with any other
- The columns in a dataframe are the observed variables, and the rows are the observations.
- You can simply display the data by executing the `df` object in a new cell


**File Not Found:**

> The name of the URL needs to be very precise, and it's case sensitive, the smallest typo can get you an error because Pandas can't locate the file you're giving

`ERROR`: FileNotFoundError: [Errno 2] No such file or directory: '..gapminder.csv'

```python
url = '..gapminder.csv'
df = pd.read_csv(url)

---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-11-5b3ae3a38afe> in <module>
      1 url = '..gapminder.csv'
----> 2 df = pd.read_csv(url)
```

## **2. Pandas basic functions**
- It is a very extensive library, some initial steps can be found in the [10 minutes to pandas tutorial](https://pandas.pydata.org/docs/user_guide/10min.html)
- Some useful tips are:

Use `df.head()` to get the first rows of the dataframe

In [2]:
df.head(3)

country continent  year  lifeExp       pop   gdpPercap iso_alpha  \
0  Afghanistan      Asia  1952   28.801   8425333  779.445314       AFG   
1  Afghanistan      Asia  1957   30.332   9240934  820.853030       AFG   
2  Afghanistan      Asia  1962   31.997  10267083  853.100710       AFG   

   iso_num  
0        4  
1        4  
2        4

Use the `df.info()` method to find out more about a dataframe

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   country    1704 non-null   object 
 1   continent  1704 non-null   object 
 2   year       1704 non-null   int64  
 3   lifeExp    1704 non-null   float64
 4   pop        1704 non-null   int64  
 5   gdpPercap  1704 non-null   float64
 6   iso_alpha  1704 non-null   object 
 7   iso_num    1704 non-null   int64  
dtypes: float64(2), int64(3), object(3)
memory usage: 106.6+ KB


* This is a DataFrame
* 1704 rows and 8 columns
* There are different data types such as float64(2), int64(3) and object(3)
* Uses 106.6+ KB bytes of memory

Use `df.columns` to know the dataframe's columns

* Note that this is an attribute (a variable associated with the dataframe), _not_ a method. (it doesn't have parentheses)
  - Like `math.pi`
  - So do not use `()` to call it

In [4]:
df.columns

Index(['country', 'continent', 'year', 'lifeExp', 'pop', 'gdpPercap',
       'iso_alpha', 'iso_num'],
      dtype='object')

Use `df.T` to transpose a dataframe:

* Sometimes want to treat columns as rows and vice versa.
* Transpose (written `.T`) doesn't copy the data, just changes the program's view of it.
* Like `columns`, it is an attribute variable


In [5]:
df.T

0            1            2            3            4     \
country    Afghanistan  Afghanistan  Afghanistan  Afghanistan  Afghanistan   
continent         Asia         Asia         Asia         Asia         Asia   
year              1952         1957         1962         1967         1972   
lifeExp         28.801       30.332       31.997        34.02       36.088   
pop            8425333      9240934     10267083     11537966     13079460   
gdpPercap      779.445      820.853      853.101      836.197      739.981   
iso_alpha          AFG          AFG          AFG          AFG          AFG   
iso_num              4            4            4            4            4   

                  5            6            7            8            9     \
country    Afghanistan  Afghanistan  Afghanistan  Afghanistan  Afghanistan   
continent         Asia         Asia         Asia         Asia         Asia   
year              1977         1982         1987         1992         1997   
lifeExp         38.438       39.854       40.822       41.674       41.763   
pop           14880372     12881816     13867957     16317921     22227415   
gdpPercap      786.113      978.011      852.396      649.341      635.341   
iso_alpha          AFG          AFG          AFG          AFG          AFG   
iso_num              4            4            4            4            4   

           ...      1694      1695      1696      1697      1698      1699  \
country    ...  Zimbabwe  Zimbabwe  Zimbabwe  Zimbabwe  Zimbabwe  Zimbabwe   
continent  ...    Africa    Africa    Africa    Africa    Africa    Africa   
year       ...      1962      1967      1972      1977      1982      1987   
lifeExp    ...    52.358    53.995    55.635    57.674    60.363    62.351   
pop        ...   4277736   4995432   5861135   6642107   7636524   9216418   
gdpPercap  ...   527.272   569.795   799.362   685.588   788.855   706.157   
iso_alpha  ...       ZWE       ZWE       ZWE       ZWE       ZWE       ZWE   
iso_num    ...       716       716       716       716       716       716   

               1700      1701      1702      1703  
country    Zimbabwe  Zimbabwe  Zimbabwe  Zimbabwe  
continent    Africa    Africa    Africa    Africa  
year           1992      1997      2002      2007  
lifeExp      60.377    46.809    39.989    43.487  
pop        10704340  11404948  11926563  12311143  
gdpPercap   693.421    792.45   672.039   469.709  
iso_alpha       ZWE       ZWE       ZWE       ZWE  
iso_num         716       716       716       716  

[8 rows x 1704 columns]

Use `df.describe()` to get summary statistics about the data

- This method returns a new samaller dataframe with only the summary statistics of the columns that have numerical data. All other columns are ignored.

In [6]:
df.describe()

year      lifeExp           pop      gdpPercap      iso_num
count  1704.00000  1704.000000  1.704000e+03    1704.000000  1704.000000
mean   1979.50000    59.474439  2.960121e+07    7215.327081   425.880282
std      17.26533    12.917107  1.061579e+08    9857.454543   248.305709
min    1952.00000    23.599000  6.001100e+04     241.165877     4.000000
25%    1965.75000    48.198000  2.793664e+06    1202.060309   208.000000
50%    1979.50000    60.712500  7.023596e+06    3531.846989   410.000000
75%    1993.25000    70.845500  1.958522e+07    9325.462346   638.000000
max    2007.00000    82.603000  1.318683e+09  113523.132900   894.000000

With this single line of code, we are able to get some statistical measures, like the Life expectancy of the world `lifeExp` is `59.47` or the max `year` available data is `2007`

Use `df.describe(include = 'object')` to get summary statistics about the categorical data

In [7]:
df.describe(include = 'object')

country continent iso_alpha
count        1704      1704      1704
unique        142         5       141
top     Venezuela    Africa       KOR
freq           12       624        24

We can also see that there are 142 countries in the dataset across 5 continents. `top` is the most frequent value (i.e., the mode) and `freq` is the frequency of the top value(number of records). In this dataset the top and frequency are not very important.

## **3. Pandas subsets and selection**
- Using Pandas it is possible to select entire rows or entire columns from a dataframe.
- Also, select a subset of a dataframe by defined criterion.
- The DataFrame is the way Pandas represents a table
- Pandas is built on top of the NumPy library, which in practice means that most of the methods defined for Numpy Arrays apply to DataFrames.

In [8]:
# import the library
import pandas as pd

# directory to the file, it can be from own computer or from web
url = 'https://raw.githubusercontent.com/MaastrichtU-IDS/global-studies/main/semester2/data/gapminder.csv'

# read the url and save it in an object called df
df = pd.read_csv(url)

We can get a slice of the data by using the `df.query()` function to select all the records relating to a specific year. For example, by specifying the corresponding column to filter (`year`) and the desired value (`2007`).

In [9]:
df.query('year == 2007')

country continent  year  lifeExp       pop     gdpPercap  \
11           Afghanistan      Asia  2007   43.828  31889923    974.580338   
23               Albania    Europe  2007   76.423   3600523   5937.029526   
35               Algeria    Africa  2007   72.301  33333216   6223.367465   
47                Angola    Africa  2007   42.731  12420476   4797.231267   
59             Argentina  Americas  2007   75.320  40301927  12779.379640   
...                  ...       ...   ...      ...       ...           ...   
1655             Vietnam      Asia  2007   74.249  85262356   2441.576404   
1667  West Bank and Gaza      Asia  2007   73.422   4018332   3025.349798   
1679         Yemen, Rep.      Asia  2007   62.698  22211743   2280.769906   
1691              Zambia    Africa  2007   42.384  11746035   1271.211593   
1703            Zimbabwe    Africa  2007   43.487  12311143    469.709298   

     iso_alpha  iso_num  
11         AFG        4  
23         ALB        8  
35         DZA       12  
47         AGO       24  
59         ARG       32  
...        ...      ...  
1655       VNM      704  
1667       PSE      275  
1679       YEM      887  
1691       ZMB      894  
1703       ZWE      716  

[142 rows x 8 columns]

**Note:** This was just a slice of the data, this new subset is not saved anywhere, in order to preserve it, we need to save it in an object. Let's define the `df_2007` object then.

In [10]:
df_2007 = df.query('year == 2007')

df_2007.head(2)

country continent  year  lifeExp       pop    gdpPercap iso_alpha  \
11  Afghanistan      Asia  2007   43.828  31889923   974.580338       AFG   
23      Albania    Europe  2007   76.423   3600523  5937.029526       ALB   

    iso_num  
11        4  
23        8

We can get yet another subset of the `df_2007` subset by using the `df.query()`. Let's filter now by a column that is not numerical, for example `continent` selecting `Europe`. Please note we are using double quotations now!

In [11]:
df_2007_europe = df_2007.query('continent == "Europe" ')

df_2007_europe.head(2)

country continent  year  lifeExp      pop     gdpPercap iso_alpha  iso_num
23  Albania    Europe  2007   76.423  3600523   5937.029526       ALB        8
83  Austria    Europe  2007   79.829  8199783  36126.492700       AUT       40

Use `DataFrame.sort_values()` to order the values

- This function expects a parameter to indicate which column you want to order by
- The second parameter indicates ascending or descending order (when ascending is `False` it's the same as descending)
- **Note:** This time we are not using `df` anymore since now we are interested only in the European subset

In [12]:
df_2007_europe.sort_values(by='pop', ascending=False)

country continent  year  lifeExp       pop     gdpPercap  \
575                  Germany    Europe  2007   79.406  82400996  32170.374420   
1583                  Turkey    Europe  2007   71.777  71158647   8458.276384   
539                   France    Europe  2007   80.657  61083916  30470.016700   
1607          United Kingdom    Europe  2007   79.425  60776238  33203.261280   
779                    Italy    Europe  2007   80.546  58147733  28569.719700   
1427                   Spain    Europe  2007   80.941  40448191  28821.063700   
1235                  Poland    Europe  2007   75.563  38518241  15389.924680   
1283                 Romania    Europe  2007   72.476  22276056  10808.475610   
1091             Netherlands    Europe  2007   79.762  16570613  36797.933320   
599                   Greece    Europe  2007   79.483  10706290  27538.411880   
1247                Portugal    Europe  2007   78.098  10642836  20509.647770   
119                  Belgium    Europe  2007   79.441  10392226  33692.605080   
407           Czech Republic    Europe  2007   76.486  10228744  22833.308510   
1343                  Serbia    Europe  2007   74.002  10150265   9786.534714   
683                  Hungary    Europe  2007   73.338   9956108  18008.944440   
1475                  Sweden    Europe  2007   80.884   9031088  33859.748350   
83                   Austria    Europe  2007   79.829   8199783  36126.492700   
1487             Switzerland    Europe  2007   81.701   7554661  37506.419070   
191                 Bulgaria    Europe  2007   73.005   7322858  10680.792820   
419                  Denmark    Europe  2007   78.332   5468120  35278.418740   
1379         Slovak Republic    Europe  2007   74.663   5447502  18678.314350   
527                  Finland    Europe  2007   79.313   5238460  33207.084400   
1151                  Norway    Europe  2007   80.196   4627926  49357.190170   
155   Bosnia and Herzegovina    Europe  2007   74.852   4552198   7446.298803   
383                  Croatia    Europe  2007   75.748   4493312  14619.222720   
755                  Ireland    Europe  2007   78.885   4109086  40675.996350   
23                   Albania    Europe  2007   76.423   3600523   5937.029526   
1391                Slovenia    Europe  2007   77.926   2009245  25768.257590   
1019              Montenegro    Europe  2007   74.543    684736   9253.896111   
695                  Iceland    Europe  2007   81.757    301931  36180.789190   

     iso_alpha  iso_num  
575        DEU      276  
1583       TUR      792  
539        FRA      250  
1607       GBR      826  
779        ITA      380  
1427       ESP      724  
1235       POL      616  
1283       ROU      642  
1091       NLD      528  
599        GRC      300  
1247       PRT      620  
119        BEL       56  
407        CZE      203  
1343       SRB      688  
683        HUN      348  
1475       SWE      752  
83         AUT       40  
1487       CHE      756  
191        BGR      100  
419        DNK      208  
1379       SVK      703  
527        FIN      246  
1151       NOR      578  
155        BIH       70  
383        HRV      191  
755        IRL      372  
23         ALB        8  
1391       SVN      705  
1019       MNE      499  
695        ISL      352

This last table looks quite neat, the following command will make it even better, by assigning the country as the index (the row names) of the dataframe

In [13]:
# european 2007 dataset sets the country as index and applies it to the current dataframe with inplace=True (with inplace=False a new dataframe is returned)
df_2007_europe.set_index('country', inplace=True )

Use `DataFrame.loc[ , ]` to select values by their (entry) label. e.g. What is the `Netherlands` population `pop`?. First the index of the row and then the column name

In [14]:
df_2007_europe.loc['Netherlands','pop']

16570613

Would produce the same result printing `df.loc["Albania"]`(without a second index).

In [15]:
df_2007_europe.loc['Albania']

continent     Europe
year            2007
lifeExp       76.423
pop          3600523
gdpPercap    5937.03
iso_alpha        ALB
iso_num            8
Name: Albania, dtype: object

The Result of slicing can be used in further operations.

- Usually don’t just print a slice.
- All the statistical operators that work on entire dataframes work the same way on slices.

Use `df.mean()` to calcluate the average of a column
- We can call a column by `df.loc[:,'pop']`
- Note that the right hand side of the index has a colon, which indicates that we want to obtain the `pop` column for all the countries

In [16]:
# select one column of the subset and assign it to a new variable
population = df.loc[:,'pop']

# calculate the average population 
print( population.mean(), 'In average in Europe in 2007' )

29601212.324530516 In average in Europe in 2007


Finally, it is possible to create a subset based on a condition, e.g. "Selecting all the countries with a life expectancy above 80 years" We save it as the variable `final_subset`

In [17]:
# Subset based on logical condition (>)
final_subset = df_2007_europe.query('lifeExp > 80')

# display it
final_subset

continent  year  lifeExp       pop    gdpPercap iso_alpha  iso_num
country                                                                       
France         Europe  2007   80.657  61083916  30470.01670       FRA      250
Iceland        Europe  2007   81.757    301931  36180.78919       ISL      352
Italy          Europe  2007   80.546  58147733  28569.71970       ITA      380
Norway         Europe  2007   80.196   4627926  49357.19017       NOR      578
Spain          Europe  2007   80.941  40448191  28821.06370       ESP      724
Sweden         Europe  2007   80.884   9031088  33859.74835       SWE      752
Switzerland    Europe  2007   81.701   7554661  37506.41907       CHE      756

**IMPORTANT**
- Having this last subset we have created: "Pop and GDP of European countries in 2007 with Life expectancy above 80"
- We can now save it in a real file, this can be CSV or Excel
- To handle Excel files an aditional library needs to be installed `xlrd`
Use `DataFrame.to_csv()` to save the file in the same directory your notebook is located

In [18]:
# In quotations is the name we WANT to give to the file
final_subset.to_csv('europe_2007_above_80.csv')

## **4. Crosstab tables with Pandas**
- Crosstab is an important function to learn
- A confusion Matrix is a type of Crosstab
- In Pandas you can create Dataframes from Vectors!

Let's define a brand new dataframe of two columns 

In [19]:
# import the library
import pandas as pd

# define two vectors as lists
vector_1 = [1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0]
vector_2 = [1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0]

# define a new dataframe with (pd.DataFrame) function
table = pd.DataFrame({'Actual':vector_1, 'Predicted': vector_2}) 

# display
table

Actual  Predicted
0        1          1
1        0          1
2        0          0
3        1          1
4        0          0
5        1          1
6        0          1
7        0          0
8        1          1
9        0          0
10       1          0
11       0          0

When calculating a confusion matrix one needs to count the number of 1's and 0's from both columns. Then aggregate them in a way that the Predicted are the columns and the Actual values are the rows. Pandas can make our life easier by using the `pd.crosstab()` function. `pd.crosstab()` function takes the first two parameters as the columns of the table


In [21]:
confusion_matrix = pd.crosstab(table['Actual'], table['Predicted'])

#display it
confusion_matrix

Predicted  0  1
Actual         
0          5  2
1          1  4

---
<i style="color:red">EXERCISES</i>

+ _1. Go to the following URL using your browser and manually download the dataset to your computer. Put the file in the same folder where your notebook is. Finally, use `pd.read_csv()` to read it.
Is there any difference in the data when reading it from the web or reading it from your computer?_

``` Python
url = 'https://raw.githubusercontent.com/MaastrichtU-IDS/global-studies/main/semester2/data/gapminder.csv'

pd.read_csv()

```
___

+ _2. Use the `help` function for `help(df.head)` and `help(df.tail)` discuss the difference and similarities_
___

+ _3. Which one of the following query functions will work and why?_

```python
df.query('year == "1992" ')

df.query('year == 1992 ')

df.query('year = "1992" ')
```
___


+ _4. Select a whole column from the `df` dataframe with the `df.loc[ , ]` method, and call it `my_column`_

_What would be the purpose of the following function?:_ 

```python
my_column.unique()
```

___

+ _5. Just like we created a subset of `Europe 2007`, create another one for `Europe` but this time for the year `1952`._  

_What is the average life expectancy for each year? How much has changed?_
___

+ _6. With the help of the command `!pip install` we can install install additional libraries to our Python environment._

_In a new cell, install the library that reads and writes Excel files_

```python
!pip install xlrd
```

Secondly, save the Confusion matrix result in an Excel file  using the Pandas command `DataFrame.to_excel()`_
```Python
table.to_excel('confusion_matrix.xls')
```
___